File: src/extract.py
import sys

from bs4 import BeautifulSoup


def extract_urls(content):
    soup = BeautifulSoup(content, 'html.parser')

    for title in soup.find_all('h3'):
        print(f'##{title.text}')
        for link in title.find_next('ul').find_all('a'):
            url = link.get('href')
            yield url


if __name__ == '__main__':
    filename = sys.argv[1]
    with open(filename) as f:
        for url in extract_urls(filename):
            print(url)

File: src/extract.rs
use scraper::ElementRef;
use scraper::Html;
use scraper::Selector;

fn extract_urls(document_content: &str) -> Vec<String> {
    let mut result = vec![];

    let html = Html::parse_document(document_content);

    let selector_h3 = Selector::parse("h3").unwrap();
    let selector_a = Selector::parse("a").unwrap();

    for elem in html.select(&selector_h3) {
        result.push(format!("##{}", elem.text().collect::<Vec<_>>().join("")).to_string());

        let ul = elem
            .next_siblings()
            .filter_map(ElementRef::wrap)
            .find(|&x| x.value().name() == "ul");

        if let Some(node) = ul {
            for elem in node.select(&selector_a) {
                if let Some(attr) = elem.value().attr("href") {
                    result.push(attr.to_string());
                }
            }
        }
    }
    result
}

fn main() {
    let args: Vec<String> = std::env::args().collect();

    let filename = args.get(1).expect("Error getting the 1st arg");
    let file = std::fs::read_to_string(filename);
    for line in extract_urls(&file.unwrap()) {
        println!("{}", line);
    }
}
